### PR TITLE
Add tests to compile generated code

### LINKED ISSUE #69: Improve test suite via language specific validations
## Description
Currently, this project relies on snapshot testing for code generation, while that is a nice approach it does not guarantee that the output code is valid in that language. We should introduce test suite that actually tried to compile the projects. 

## Additional context
We use prettier for some projects, which does this to some extent (before snapshot i.e. dev time) but proper solution is to run compile commands of various languages in CI.

### PR DESCRIPTION
**Pull request by Weakky**
Tries to compile every tests to make sure there are no errors

- [x] Typescript
- [x] Flow

Fixes #69

### CODE REVIEW

**Comment by jasonkuhrt** [comment]
Started onboarding myself into this pr tonight. A small dep change allowed the tests to pass locally. They pass on ci too, but then (it seems) jest never exits and so the ci times out. Aside from ramping up, a difficulty here is the lack of feedback. To help narrow the issue, I may try to incrementally temporally eliminate test code until the ci problem goes away.

**Comment by jasonkuhrt** [comment]
I didn't observe carefully before. Looking again at the tests running on https://circleci.com/gh/prisma/graphqlgen/677, I noticed that large-schema tests are missing from the output, e.g. from local:

```
 PASS  src/tests/flow/large-schema.test.ts (8.686s)
 PASS  src/tests/typescript/large-schema.test.ts (13.524s)
```

File: packages/graphqlgen/package.json [hunk 1]
```diff
@@ -17,7 +17,7 @@
     "build": "npm run lint && tsc --declaration",
     "watch": "tsc -w",
     "lint": "tslint --project tsconfig.json {src,test}/**/*.ts",
-    "t": "jest",
+    "t": "jest 'large schema'",
```
Comment (jasonkuhrt) [inline packages/graphqlgen/package.json]: Revert

**Comment by jasonkuhrt** [comment]
When I isolate the `large-schema` tests they run ok though [build link](https://circleci.com/gh/prisma/graphqlgen/679).

**Comment by jasonkuhrt** [comment]
`jest --runInBand` is one solution. It increases the test time however. Will continue looking for a root cause.

**Comment by jasonkuhrt** [comment]
When I comment out:

```js
  if (config.language === 'flow') {
    await compileFlow(config.models.files!, config.output)
  }
```

it works.

Looking at the source:

```js
writeFileSync(path.join(path.dirname(typesPath), '.flowconfig'), flowConfig)
```

we see disk interaction. Since tests run in parallel, and the file name here is a constant, there is disk contention. And indeed if we remove one of the schema tests (only tried "basic") of flow, the test suite works ([build link](https://circleci.com/gh/prisma/graphqlgen/686)).

I will look to see if flow has a better programatic API to work with. Writing a config to disk etc. feels like a more complicated program, is not as easily isolated as with just in memory, ... so on.

**Comment by jasonkuhrt** [comment]
Related:

* https://discuss.circleci.com/t/babel-7-jest-results-in-timeout/25383/5
* https://vgpena.github.io/jest-circleci/
* https://circleci.com/ideas/?idea=CCI-I-578

**Comment by jasonkuhrt** [comment]
Eventually figured out that the disk contention was not the root cause (probably not implicated at all). There are quite a few existing issues on the inerwebs about cci/travis not handling jest auto-cpu-count detection well. Capping worker count on CI worked here too.

I found one flow test that was commented out and needed finishing, `default name`. What I found was that flow wouldn't compile because (not type name mismatch):

```ts
export interface Mutation_AddMemberData {
  email: string;
  projects: string[];
}

export interface Mutation_Args_AddMember {
  data: AddMemberData;
}
```

I now append prefixes to the types of field arguments (in the above case... `Mutation_`) to comply with the pattern in our snapshots. I do not _really_ understand the flow strategy, so, I'm not 100% confident with my commit here; in a nutshell: every field arg type gets a prefix except enum, input, object (which correspond to at least in some tests to GQL `ID` type).

File: packages/graphqlgen/src/generators/flow-generator.ts [hunk 2]
```diff
@@ -2,7 +2,11 @@ import * as os from 'os'
 import * as prettier from 'prettier'
```
Comment (jasonkuhrt) [inline packages/graphqlgen/src/generators/flow-generator.ts]: This module contains bug-fix code in response to the new compilation step we do.

File: packages/graphqlgen/src/tests/generation.ts [hunk 3]
```diff
@@ -1,14 +1,141 @@
...
+        resolve(stdout)
+      }
+    })
+  })
+}
+
+function printTypescriptErrors(diagnotics: ReadonlyArray<ts.Diagnostic>) {
+  diagnotics.forEach(diagnostic => {
+    if (diagnostic.file) {
+      let { line, character } = diagnostic.file.getLineAndCharacterOfPosition(
```
Comment (timsuchanek) [inline packages/graphqlgen/src/tests/generation.ts]: Is there a reason to not use `const` here?

File: packages/graphqlgen/src/tests/generation.ts [hunk 4]
```diff
@@ -1,14 +1,141 @@
...
+        diagnostic.start!,
+      )
+      let message = ts.flattenDiagnosticMessageText(
```
Comment (timsuchanek) [inline packages/graphqlgen/src/tests/generation.ts]: Same here

**Comment by jasonkuhrt** [comment]
@timsuchanek let me know if there are any other blocking issues in this PR from your point of view. Propose that minor issues we punt for iterative refactoring : )

**Comment by jasonkuhrt** [comment]
@timsuchanek given that you seemed pretty satisfied already and I've fixed the two style issues you noted, going to merge, keep things rolling. And these tests like you said will start helping immediately : )

**Pull request merged**